prusti-interface/src/specs/encoder.rs

```diff
@@ -59,47 +79,76 @@ impl<'tcx> Encoder for DefSpecsEncoder<'tcx> {
-        s.tcx.stable_crate_id(*self).encode(s)
-    }
-}
-
-// See https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/rmeta/encoder/struct.EncodeContext.html
-impl<'tcx> Encodable<DefSpecsEncoder<'tcx>> for Span {
-    fn encode(&self, s: &mut DefSpecsEncoder<'tcx>) {
-        let sm = s.tcx.sess.source_map();
-        for bp in [self.lo(), self.hi()] {
+impl<'a, 'tcx> SpanEncoder for DefSpecsEncoder<'a, 'tcx> {
+    fn encode_span(&mut self, span: Span) {
+        let sm = self.tcx.sess.source_map();
+        let local_crate_stable_id = self.tcx.stable_crate_id(LOCAL_CRATE);
+        for bp in [span.lo(), span.hi()] {
             let sf = sm.lookup_source_file(bp);
-            let ssfi = StableSourceFileId::new(&sf);
-            ssfi.encode(s);
+
+            let ssfi =
+                StableSourceFileId::from_filename_for_export(&sf.name, local_crate_stable_id);
+            ssfi.encode(self);
             // Not sure if this is the most stable way to encode a BytePos. If it fails
             // try finding a function in `SourceMap` or `SourceFile` instead. E.g. the
             // `bytepos_to_file_charpos` fn which returns `CharPos` (though there is
             // currently no fn mapping back to `BytePos` for decode)
-            (bp - sf.start_pos).encode(s);
+            (bp - sf.start_pos).encode(self);
         }
     }
+    fn encode_symbol(&mut self, sym: Symbol) {
+        // // if symbol preinterned, emit tag and symbol index
+        // if sym.is_preinterned() {
+        //     self.opaque.emit_u8(SYMBOL_PREINTERNED);
+        //     self.opaque.emit_u32(sym.as_u32());
+        // } else {
+        //     // otherwise write it as string or as offset to it
+        //     match self.symbol_table.entry(sym) {
+        //         Entry::Vacant(o) => {
+        //             self.opaque.emit_u8(SYMBOL_STR);
+        //             let pos = self.opaque.position();
+        //             o.insert(pos);
+        //             self.emit_str(sym.as_str());
+        //         }
+        //         Entry::Occupied(o) => {
+        //             let x = *o.get();
+        //             self.emit_u8(SYMBOL_OFFSET);
+        //             self.emit_usize(x);
+        //         }
+        //     }
+        // }
+        todo!()
+    }
+
+    fn encode_expn_id(&mut self, eid: ExpnId) {
+        self.hygiene_context.schedule_expn_data_for_encoding(eid);
+        eid.krate.encode(self);
+        eid.local_id.as_u32().encode(self);
+    }
+    fn encode_syntax_context(&mut self, ctx: SyntaxContext) {
+        raw_encode_syntax_context(ctx, self.hygiene_context, self);
+    }
+    fn encode_crate_num(&mut self, cnum: CrateNum) {
+        self.tcx.stable_crate_id(cnum).encode(self)
+    }
+    fn encode_def_index(&mut self, _: DefIndex) {
+        panic!("encoding `DefIndex` without context");
+    }
+    fn encode_def_id(&mut self, id: DefId) {
+        self.tcx.def_path_hash(id).encode(self)
+    }
```

Comment:
@Aurel300 If you're interested, in Creusot we have all this logic extracted in a generic, creusot independent manner in a standalone crate. we could move that to a separate repo and reuse it in Prusti directly. This stuff is annoying and actually rather tricky to get 100% correct.

Reply:
Sounds good, which crate is it?

Reply:
It's called `creusot-metadata` in our repo currently